Publish our INT32, UINT32, INT64, UINT64 ASN.1 types and Z variants

levitte opened the pull request.

levitte committed:
Publish our INT32, UINT32, INT64, UINT64 ASN.1 types and Z variants

levitte committed:
Discourage the use of LONG and ZLONG, and deprecate it in the future

levitte commented:
Note that this includes a proposed future deprecation of LONG and ZLONG

mattcaswell commented on include/openssl/asn1t.h:
```diff
@@ -901,8 +901,24 @@ DECLARE_ASN1_ITEM(ASN1_FBOOLEAN)
+DECLARE_ASN1_ITEM(UINT64)
+DECLARE_ASN1_ITEM(ZUINT64)
+
+# if OPENSSL_API_COMPAT < 0x10200000L
```
So this deprecates from version 1.2.0 onward. I am unclear what our policy is for deprecating in binary compatible releases. Are we not allowed to do it?

levitte replied:
Deprecating in 1.1.*x* takes away part of the public API, and that contrary to what we say [here](https://www.openssl.org/docs/faq.html#MISC8).

mattcaswell commented:
If we are deprecating LONG/ZLONG shouldn't we also remove all uses of it in our own code? I notice there are "LONG" uses in the ec, cms and dsa code, as well as ZLONG use in the dh code. In looking for that I also fell upon asn1_item_list.h which looks like it needs an update.

levitte commented:
> If we are deprecating LONG/ZLONG shouldn't we also remove all uses of it in our own code? I notice there are "LONG" uses in the ec, cms and dsa code, as well as ZLONG use in the dh code. In looking for that I also fell upon asn1_item_list.h which looks like it needs an update.

Yup, I'm working through those...

levitte committed:
Act on deprecation of LONG and ZLONG, step 1

Don't compile code that still uses LONG when it's deprecated

levitte committed:
Act on deprecation of LONG and ZLONG, step 2

Replace all remaining uses of LONG and ZLONG with INT32 / ZINT32.

levitte commented:
There.

levitte commented:
To experiment with this, apply this patch:

``` perl
diff --git a/Configure b/Configure
index 66541be41d..9b425a4c47 100755
--- a/Configure
+++ b/Configure
@@ -171,6 +171,7 @@ our $BSDthreads="-pthread -D_THREAD_SAFE -D_REENTRANT";
 #
 my $maxapi = "1.1.0";           # API for "no-deprecated" builds
 my $apitable = {
+    "1.2.0" => "0x10200000L",
     "1.1.0" => "0x10100000L",
     "1.0.0" => "0x10000000L",
     "0.9.8" => "0x00908000L",
```

then configure with `--api=1.2.0`

levitte commented:
Note, however, that it will fail to build until #3127 has been applied...

richsalz commented:
Let’s discuss this on the list?

levitte commented:
Which part?  'cause there are two parts, one is simply publishing, the other is deprecating the old.  Maybe I should split this up in two PRs to avoid confusion?

levitte commented:
@richsalz, I suggest you start the discussion.  I have no idea what topic you want to talk about.

levitte committed:
Act on deprecation of LONG and ZLONG, step 3

Teach util/mkdef.pl to recognise these lines:

    #if OPENSSL_API_COMPAT < 0xXXXXXXXXL

    #if OPENSSL_API_COMPAT >= 0xXXXXXXXXL

and add corresponding markers in util/*.num

A final 'make update' sets those markers right for LONG and ZLONG.

levitte committed:
[to be squashed] also do not compile the LONG and ZLONG code

kaduk commented on crypto/asn1/x_long.c:
```diff
@@ -12,6 +12,9 @@
 #include <openssl/asn1t.h>
 #include <openssl/bn.h>
 
+#if !(OPENSSL_API_COMPAT < 0x10200000L)
```
Is the idea that OPENSSL_API_COMPAT will always be used with the '<' operator?  (Should that get written down somewhere?)

levitte replied:
There places where we use >=. I did it this way so the expression is kept similar to the one in asn1t.h...

levitte committed:
Add a note in CHANGES

levitte commented:
Ping?

levitte commented:
Oh, a Travis failure...

levitte committed:
Fix util/mkdef.pl

The deprecation checking code here didn't work the same way as in
Configure, and used $config{options} to find an --api= option that
was never there.  This is replaced with checking $config{api}, which
is the controlling variable for deprecation.

levitte commented:
Ping?

richsalz reviewed (commented):
Looks good to me, but perhaps @snhenson should review?  Otherwise +1 from me.

levitte commented:
> Looks good to me, but perhaps @snhenson should review? Otherwise +1 from me.

I would appreciate such a review, but will not wait endlessly.

mattcaswell reviewed (approved):
LGTM

levitte commented:
Merged.  64f11ee88883af8995d2991307ff3fb560a96fa6 to f120fa1efe6377850d1ecd389a02e0e2241912bc
